text/0000-dyno.md

```diff
@@ -0,0 +1,294 @@
+
+The `TypeTag` trait is an abstraction over all type tags. It does not have any methods, only an associated type for the type which the tag represents. E.g., `<Ref<T> as TypeTag<'a>>::Type` is `&'a T`.
+
+There is no universal type tag. A concrete type tag must be written for a 'category' of types. A few common tags are provided in `provide_any::tags`, including `Value` for any type bounded by `'static`, and `Ref` for types of the form `&'a T` where `T: 'static`. For less common types, the user must provide a tag which implements `TypeTag`; in this way the `provide_any` API generalises to all types.
+
+Intermediate libraries (such as `error` in our examples) can choose to support only common type patterns so that the user is totally unaware of type tags, or a general API using type tags, or a hybrid approach where common cases are explicitly supported for ease of use and type tags are also supported for completeness.
+
+Note, that in the specialised variations, the user can use the type directly, e.g., `get_context_ref::<Backtrace>()`. But in the general case the user must provide the type tag, not the type itself, e.g., `get_by_tag::<tags::Ref<Backtrace>>()`. That prevents using type inference or type ascription for access, and the type tag must always be explicitly specified.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+For details of the implementation, see the [provide-any](https://github.com/nrc/provide-any) repo which has a complete (though unpolished) implementation of this proposal, and an implementation of the extensions to `Error` used in the examples above.
+
+I propose that `provide_any` include the following type tags:
+
+```rust
+pub mod provide_any {
+    pub mod tags {
+        use super::TypeTag;
+        use core::marker::PhantomData;
+
+        /// Type-based `TypeTag` for `&'a T` types.
+        pub struct Ref<T: ?Sized + 'static>(PhantomData<T>);
+
+        impl<'a, T: ?Sized + 'static> TypeTag<'a> for Ref<T> {
+            type Type = &'a T;
+        }
+
+        /// Type-based `TypeTag` for `&'a mut T` types.
+        pub struct RefMut<T: ?Sized + 'static>(PhantomData<T>);
+
+        impl<'a, T: ?Sized + 'static> TypeTag<'a> for RefMut<T> {
+            type Type = &'a mut T;
+        }
+
+        /// Type-based `TypeTag` for static `T` types.
+        pub struct Value<T: 'static>(PhantomData<T>);
+
+        impl<'a, T: 'static> TypeTag<'a> for Value<T> {
+            type Type = T;
+        }
+
+        /// Tag combinator to wrap the given tag's value in an `Option<T>`
+        pub struct Optional<I>(PhantomData<I>);
```

Comment:
Along with the `Optional` combinator, can we have a `Resultant` combinator?

It seems weird to support `Option`-wrapping, but not `Result`-wrapping. And it would also help illustrate the expressive power of the type tag design.

Reply:
The reason for having Optional but not resultant is that Option is used in the implementation and is thus required. There is nothing stopping us adding Result too, I believe.